runtime/hashing/common/src/aws/smithy/kotlin/runtime/hashing/HashingSink.kt

```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.smithy.kotlin.runtime.hashing
+
+import aws.smithy.kotlin.runtime.io.SdkSink
+import aws.smithy.kotlin.runtime.io.internal.SdkSinkObserver
+
+/**
+ * A sink which hashes data as it is being written to
+ * @param hash The [HashFunction] to hash data with
+ * @param sink the [SdkSink] to hash
+ */
+public class HashingSink(private val hash: HashFunction, sink: SdkSink) : SdkSinkObserver(sink) {
+    override fun observe(data: ByteArray, offset: Int, length: Int) {
+        hash.update(data, offset, length)
+    }
+
+    /**
+     * Provides the digest value as an unsigned integer for the CRC32 function family.
+     * @return unsigned integer representing the value of the digest, if the [HashFunction] is [Crc32] or [Crc32c], and null otherwise.
+     */
+    public fun digestValue() : UInt? = (hash as Crc32?)?.digestValue()
```

Comment:
I think either special case OR move this to an `internal` package as an extension function (would require making `hash` `internal` rather than `private`